### PR TITLE
Add "Room Panel" feature

### LINKED ISSUE #238: Proposal: new panel keeping track of mobs and players present in the current room
I am proposing a new feature, and if accepted I am also offering to implement it myself:

Add a new Mob/Player Panel, somewhat similar to the already existing Group Panel, that shows the list of mobs and players present in the current room.
The idea is to keep it up-to-date to the best of MMapper abilities, i.e. updating it when:
- the player moves
- some mob or player enter or leaves the room, provided that the player can see it (i.e. if MUME sends the message "Foo leaves north/south/east/west/up/down")

And also taking into account that movement of other mobs/players in the same [group](https://mume.org/help/group) as the player are omitted by MUME in many cases - for example the leader of a group receives a message like
```ACK! A pack horse didn't follow you, you lost him.```
when he/she loses a follower, but in nospam mode MUME will **not** tell if a follower successfully follows his/her leader.

Another useful feature could be:
- showing in the mob/player list what **each** one is doing, who he/she is attacking etc.

i.e. the panel is basically intended to be a snapshot of `look` or `examine`, and also updated in real-time to the best of MMapper abilities when the player or somebody else moves, or someone in the room does something significant (attacking, casting, resting, standing etc).

I am interested in feedback for the proposal above, in particular:
- does it fit MMapper features and functions, or it would be better implemented in some other program?
- does it sound useful to MUME players, or it would need some minor/major changes to be useful?

My hope is that such feature could be **useful in situations with a lot of mobs/players** - **including PK situations** - where **messages scroll very quickly and it's difficult to keep track** of who is in the room and what they are doing.

### PR DESCRIPTION
**Pull request by cosmos72**
Fixes #238 - implement "Room Panel" showing GMCP messages Room.Chars.* i.e. show in a new dedicated panel
the list of mobs/players in current room and who they are fighting

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -352,6 +352,14 @@ set(mmapper_SRCS
     proxy/proxy.h
     proxy/telnetfilter.cpp
     proxy/telnetfilter.h
+    roompanel/CRoom.cpp
```
Comment (nschimme) [inline src/CMakeLists.txt:355]: The prefix of "C" is a PandoraMapper naming anti-pattern. We're starting to converge on CamelCase naming convention for files

File: src/roompanel/roomwidget.cpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        const int row = index.row();
+        const int column = index.column();
+
+        switch (role) {
+        case Qt::DisplayRole:
+            return getMobField(row, column);
+        case Qt::BackgroundRole: {
+            const QColor color = getMobColor(row);
+            if (color.isValid()) {
+                // honor CRoomMob color - hilights enemies
```
Comment (nschimme) [inline src/roompanel/roomwidget.cpp:91]: highlight :-)

File: src/CMakeLists.txt (hunk 1 above)
Comment (cosmos72) [reply src/CMakeLists.txt]: fixed

File: src/roompanel/roomwidget.cpp (hunk 2 above)
Comment (cosmos72) [reply src/roompanel/roomwidget.cpp]: fixed too :-)

File: src/pandoragroup/mmapper2group.cpp [hunk 3]
```diff
@@ -564,7 +564,7 @@ void Mmapper2Group::slot_parseGmcpInput(const GmcpMessage &msg)
 
     if (msg.isCharVitals()) {
         // "Char.Vitals {\"hp\":100,\"maxhp\":100,\"mana\":100,\"maxmana\":100,\"mp\":139,\"maxmp\":139}"
-        QJsonDocument doc = QJsonDocument::fromJson(msg.getJson()->toQString().toUtf8());
+        const QJsonDocument &doc = msg.getDocument();
```
Comment (nschimme) [inline src/pandoragroup/mmapper2group.cpp:567]: Is there a reason we decided to pre-parse the JSON? The main reason I did it in a lazy fashion was because MMapper doesn't care about all GMCP messages and we could avoid some deserialization overhead.

Also, if MMapper only deserializes the message once then we aren't really gaining anything by doing it ahead of time. I assume this is the case here.

File: src/display/mapcanvas.cpp [hunk 4]
```diff
@@ -1077,7 +1077,7 @@ void MapCanvas::userPressedEscape(bool /*pressed*/)
 
     case CanvasMouseModeEnum::MOVE:
         // special case for move: right click selects infomarks
-        FALLTHRU;
+        FALLTHROUGH;
```
Comment (nschimme) [inline src/display/mapcanvas.cpp:1080]: We should extract this out and package it as a separate PR enhancement

File: src/global/Flags.h [hunk 5]
```diff
@@ -179,23 +179,23 @@ class NODISCARD Flags
     }
 
 public:
-    NODISCARD inline bool contains(const Flag flag) const noexcept
+    NODISCARD inline constexpr bool contains(const Flag flag) const noexcept
```
Comment (nschimme) [inline src/global/Flags.h:182]: These constexpr enhancements should also be a separate PR

File: src/pandoragroup/CGroupChar.cpp [hunk 6]
```diff
@@ -50,25 +51,6 @@ const QVariantMap CGroupChar::toVariantMap() const
     return root;
 }
 
-// TODO: move this somewhere more appropriate
```
Comment (nschimme) [inline src/pandoragroup/CGroupChar.cpp:53]: Thanks!

File: src/proxy/GmcpMessage.h [hunk 7]
```diff
@@ -62,9 +66,14 @@ using GmcpJson = TaggedStringUtf8<tags::GmcpJsonTag>;
 class GmcpMessage final
 {
 private:
-    GmcpMessageName name;
-    std::optional<GmcpJson> json;
```
Comment (nschimme) [inline src/proxy/GmcpMessage.h:66]: Why give up on std::optional? Doing an empty check is basically the same thing but it's less obvious to the API user that its actually an optional payload according to the GMCP spec.

File: src/roompanel/RoomManager.cpp [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+#include <utility>
+#include <QtCore>
+
+#include "../global/StringView.h"
+#include "../proxy/GmcpMessage.h"
+#include "RoomManager.h"
+#include "RoomMobs.h"
+
+RoomManager::RoomManager(QObject *const /*parent*/)
+    : QObject{nullptr}
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:16]: Why not pass in the QMainWindow as the parent? This is will otherwise be leaked when the application is closed.

File: src/roompanel/RoomManager.cpp [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    , m_room{std::make_unique<RoomMobs>(nullptr)}
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:17]: Why not make it here? The parent can be the RoomManager ?

File: src/roompanel/RoomManager.cpp [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    addMob(doc.object());
+}
+
+void RoomManager::parseGmcpRemove(const GmcpMessage &msg)
+{
+    showGmcp(msg);
+    // payload is a single number (usually followed by a space), not an array or object
+    // thus parsed msg.getDocument() is not valid: use msg.getJson() instead
+    const std::string &str = msg.getJson().getStdString();
+    const unsigned long num = std::stoul(str);
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:76]: Would doing some bound checks here help anything?

File: src/roompanel/RoomManager.cpp [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    const RoomMob::Id id = static_cast<RoomMob::Id>(num);
+    if (id == 0 || static_cast<unsigned long>(id) != num) {
+        if (false) {
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:79]: Why disable this warning? Wouldn't we want to discover this error?

File: src/roompanel/RoomManager.cpp [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+void RoomManager::parseGmcpSet(const GmcpMessage &msg)
+{
+    showGmcp(msg);
+    if (isInvalidJson(msg)) {
+        return;
+    }
+    const QJsonDocument &doc = msg.getDocument();
+    if (!doc.isArray()) {
+        if (false) {
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:99]: A member variable (i.e. `debug`) that we can toggle false/true will help us having to manually replace each of these if statements when debugging.

File: src/roompanel/RoomManager.cpp [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+{
+    QJsonObject::iterator iter = obj.find("id");
+    QJsonObject::iterator end = obj.end();
+    if (iter == end || !toMobId(*iter, data.id())) {
+        return false;
+    }
+    auto matchEnd = mobFields.end();
+    for (iter = obj.begin(); iter != end; ++iter) {
+        auto match = mobFields.find(iter.key());
+        if (match != matchEnd) {
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:200]: What's the impact of this being matchEnd? Should we handle this?

File: src/roompanel/RoomManager.cpp [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    return true;
+}
+
+bool RoomManager::toMobId(QJsonValueRef value, RoomMob::Id &id)
+{
+    // RoomMob::Id cannot represent negative numbers
+    const double d = value.toDouble(-1.0);
+    id = static_cast<RoomMob::Id>(d);
+    // check for exact conversion
+    return std::equal_to<double>{}(d, static_cast<double>(id));
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:213]: This isn't very standard... let's pick one of these exceptions: https://www.cplusplus.com/reference/stdexcept/

File: src/roompanel/RoomManager.cpp [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+    switch (value.type()) {
+    case QJsonValue::Double:
+        variant = static_cast<unsigned>(value.toDouble());
+        break;
+    case QJsonValue::String:
+        str = value.toString();
+        if (str == RoomMobData::you) {
+            str = RoomMobData::YOU;
+        }
+        if (!str.isEmpty() && str.at(0) == QChar('*')) {
```
Comment (nschimme) [inline src/roompanel/RoomManager.cpp:230]: Please add a comment for why `*`. I imagine this is for `*an Orc*` but it's not super clear.

File: src/roompanel/RoomWidget.cpp [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        case ColumnTypeEnum::EFFECTS:
+            return "Effects";
+        case ColumnTypeEnum::WEAPON:
+            return "Weapon";
+        case ColumnTypeEnum::FIGHTING:
+            return "Fighting";
+        case ColumnTypeEnum::MOUNT:
+            return "Mount";
+        default:
+            if (false) {
```
Comment (nschimme) [inline src/roompanel/RoomWidget.cpp:64]: Why not warn here?

File: src/roompanel/RoomWidget.cpp [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+        return mob->color();
+    }
+    return QColor{};
+}
+
+bool RoomModel::isFightingYOU(const int row, const int column) const
+{
+    if (static_cast<ColumnTypeEnum>(column) == ColumnTypeEnum::FIGHTING //
+        && row >= 0 && size_t(row) < m_mobVector.size()) {
+        //
```
Comment (nschimme) [inline src/roompanel/RoomWidget.cpp:184]: Why these dangling comments?

File: src/roompanel/RoomWidget.cpp [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+{
+    beginResetModel();
+    m_room.listMobs(m_mobsById, m_mobVector);
+    endResetModel();
+}
+
+// ------------------------------- RoomWidget ----------------------------------
+RoomWidget::RoomWidget(RoomManager *rm, QWidget *const parent)
+    : QWidget{parent}
+    , m_model{nullptr}
```
Comment (nschimme) [inline src/roompanel/RoomWidget.cpp:201]: Don't initialize here then since we're doing it below

File: src/roompanel/RoomWidget.cpp [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+    m_model = std::make_unique<RoomModel>(this, *rm->getRoom());
+
+    m_table = std::make_unique<QTableView>(this);
+    m_table->setSelectionMode(QAbstractItemView::ContiguousSelection);
+    m_table->setSelectionBehavior(QAbstractItemView::SelectRows);
+    m_table->horizontalHeader()->setStretchLastSection(false);
+    m_table->horizontalHeader()->setSectionResizeMode(QHeaderView::Interactive);
+    m_table->setModel(m_model.get());
+    /// TODO: [Cosmos] m_table->setItemDelegate(new RoomDelegate(this));
```
Comment (nschimme) [inline src/roompanel/RoomWidget.cpp:218]: Care to make this comment be more human readable?

File: src/roompanel/RoomWidget.h [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+signals:
+
+private:
+    void readSettings();
+    void writeSettings();
+
+private:
+    std::unique_ptr<RoomModel> m_model;
+    std::unique_ptr<QTableView> m_table;
+    RoomManager *m_roomManager;
```
Comment (nschimme) [inline src/roompanel/RoomWidget.h:82]: Could we pass it in by reference instead?

**Review by nschimme — CHANGES_REQUESTED** [review]

**Comment by nschimme** [comment]
Any chance we could also write some unit tests for this?

**Comment by nschimme** [comment]
@cosmos72 oops, I think I just force pushed over you. Please rebase on top of my change, sorry!

File: src/pandoragroup/mmapper2group.cpp (hunk 3 above)
Comment (cosmos72) [reply src/pandoragroup/mmapper2group.cpp:567]: The `GmcpMessage` class originally (i.e. before my changes) only contained unparsed JSON inside an UTF-8 string,
and parsed it every time it needed the JSON.

I first changed to it lazily parse the UTF-8 string to JSON on demand, but then I thought about multi-thread synchronization issues - i.e. thread safety - and finally I decided to just parse the UTF-8 string to JSON when GmcpMessage is created.

If you think parsing UTF-8 to JSON is a potentially expensive operation, and it's worth skipping it for messages MMapper is not interested into, I will put back the lazy parsing and its accompanying synchronization.

File: src/display/mapcanvas.cpp (hunk 4 above)
Comment (cosmos72) [reply src/display/mapcanvas.cpp:1080]: I could not resist fixing the typo, er, abbreviation :)

File: src/roompanel/RoomManager.cpp (hunk 8 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:16]: I'm not an expert of Qt objects lifecycle... I will fix this.

File: src/roompanel/RoomManager.cpp (hunk 10 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:76]: Yes, we can ignore strings that contain trail junk or that parse to out-of-range integers. I'll add that.

File: src/roompanel/RoomWidget.cpp (hunk 17 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.cpp:185]: the `//` are there to force IDEs or clang-format to split these long lines

File: src/roompanel/RoomWidget.cpp (hunk 19 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.cpp:219]: This is a leftover of copy-pasting `GroupPanel` code. To be honest, I have no idea what's an `ItemDelegate` is. Any reference?

**Comment by cosmos72** [comment]
This branch exists for the pull request, no problem if you pushed commits to it.

File: src/roompanel/RoomWidget.cpp (hunk 19 above)
Comment (nschimme) [reply src/roompanel/RoomWidget.cpp:219]: QStyledItemDelegate allows you to replace the text with some icon: https://doc.qt.io/qt-5/qtwidgets-itemviews-stardelegate-example.html

We used it in the GroupPanel to change the text to little icons

File: src/global/Flags.h (hunk 5 above)
Comment (nschimme) [reply src/global/Flags.h]: I did this

File: src/pandoragroup/mmapper2group.cpp (hunk 3 above)
Comment (cosmos72) [reply src/pandoragroup/mmapper2group.cpp:567]: [UPDATE] I modified `GmcpMessage` to lazily parse string to JSON

File: src/roompanel/RoomWidget.cpp (hunk 19 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.cpp]: I see. Removed, not using icons in RoomPanel at the moment.

File: src/roompanel/RoomWidget.h (hunk 20 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.h]: Done

File: src/roompanel/RoomWidget.cpp (hunk 18 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.cpp]: Done.

File: src/roompanel/RoomManager.cpp (hunk 8 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp]: Fixed

File: src/roompanel/RoomManager.cpp (hunk 9 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp]: Fixed. I set RoomMobs's parent to `parent` i.e. the main window

File: src/roompanel/RoomManager.cpp (hunk 13 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:200]: All JSON object keys are optional in GMCP messages Room.Chars.*  - only 'id' is mandatory.

If `match == matchEnd` it means one of the JSON object keys we know about (name, desc, fighting, flags, labels, riding, driving, position, weapon) is not present in the GMCP message, so we don't save the corresponding value into `RoomMobUpdate`

File: src/display/mapcanvas.cpp (hunk 4 above)
Comment (cosmos72) [reply src/display/mapcanvas.cpp]: Reverted. The rename `FALLTHRU` -> `FALLTHROUGH` can be done in a different PR

File: src/global/Flags.h (hunk 5 above)
Comment (cosmos72) [reply src/global/Flags.h]: shall I revert these constexpr enhancements?

File: src/roompanel/RoomManager.cpp (hunk 10 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp]: Added.

File: src/roompanel/RoomManager.cpp (hunk 12 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp]: Done.

File: src/roompanel/RoomManager.cpp (hunk 13 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:215]: I have added
```C++
if (m_debug) {
  qWarning() << // ...
}
```

File: src/roompanel/RoomManager.cpp (hunk 14 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:231]: Throw an exception just to immediately catch it in the caller, swallow it and return?
It really seems unnecessarily complicated (and slow) to me.

I wanted to write `return d == static_cast<double>(id);` but some compilers complained that `operator==` on double shouldn't be used.

File: src/roompanel/RoomManager.cpp (hunk 15 above)
Comment (cosmos72) [reply src/roompanel/RoomManager.cpp:249]: Exactly, it's to highlight enemy players. I added a comment.

File: src/roompanel/RoomWidget.cpp (hunk 16 above)
Comment (cosmos72) [reply src/roompanel/RoomWidget.cpp]: I have replaced it with
```C++
if (m_debug) {
  qWarning() << qWarning() << "Unsupported column" << column;
}
```

File: external/zlib/CMakeLists.txt [hunk 21]
```diff
@@ -33,8 +33,8 @@ else()
     message(STATUS "Could not find system zlib; building as a static library")
     include(ExternalProject)
     ExternalProject_Add(zlib
-        URL "https://zlib.net/zlib-1.2.11.tar.gz"
-        URL_HASH SHA256=c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1
+        URL "https://zlib.net/zlib-1.2.12.tar.gz"
+        URL_HASH SHA256=91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9
```
Comment (nschimme) [inline external/zlib/CMakeLists.txt:37]: Thanks! This should be a separate PR so we can merge it sooner!

Comment (cosmos72) [reply external/zlib/CMakeLists.txt:37]: Of course :)
Does it make sense to cherry-pick it for merging in master branch?

**Comment by cosmos72** [comment]
I have rebased 'roompanel' branch - and as a consequence, this PR - on master commit [487b64d0069f3a99e86e04321e106fbe5b0f8c76](https://github.com/MUME/MMapper/commit/487b64d0069f3a99e86e04321e106fbe5b0f8c76)

Hopefully I did not do anything plainly wrong - it's the first time I used `git rebase --interactive` and I also tried to meld together related commits, as per your suggestion

**Comment by nschimme** [comment]
> I have rebased 'roompanel' branch - and as a consequence, this PR - on master commit [487b64d0069f3a99e86e04321e106fbe5b0f8c76](https://github.com/MUME/MMapper/commit/487b64d0069f3a99e86e04321e106fbe5b0f8c76)
> 
> Hopefully I did not do anything plainly wrong - it's the first time I used `git rebase --interactive` and I also tried to meld together related commits, as per your suggestion

I took a stab at organizing the commits a little further.
